libtenzir/src/node.cpp

```diff
@@ -284,6 +284,48 @@ auto node_state::get_endpoint_handler(const http_request_description& desc)
   return result->second;
 }
 
+auto spawn_components(node_actor::stateful_pointer<node_state> self) -> void {
+  using component_plugin_map = std::unordered_map<std::string, const component_plugin*>;
+  component_plugin_map loaded = {};
+  component_plugin_map not_loaded = {};
+  for (const auto* component : plugins::get<component_plugin>()) {
+    not_loaded.emplace(component->component_name(), component);
+  }
+  auto load_recursive = [&](auto load_recursive, const std::string& needed) {
+    if (loaded.contains(needed)) {
+      return;
+    }
+    auto entry = not_loaded.find(needed);
+    if (entry == not_loaded.end()) {
+      TENZIR_WARN("{} is trying to load the unknown component {}, expect "
+                  "reduced functionality.",
+                  *self, needed);
+      return;
+    }
+    for (auto& required : entry->second->required_components()) {
+      load_recursive(load_recursive, required);
+    }
+    auto handle = entry->second->make_component(self);
```

Comment:
The shutdown order is currently untouched. This is based on the assumption that these dependencies are only relevant during component initialization (for example for starting configured pipelines) but have no effect for the shutdown, but this might not be right.

Imho the correct way to model the shutdown of components would be to keep a data structure that models the dependency graph in the node and holds actor handles to the components to increase the ref-count. Once the shutdown is triggered the components that are not referenced by others in the graph can be removed, the rest will be removed step by step in the down_handler when they don't have any live dependencies any more.

Reply:
The simple solution would be to keep the sequential loading order and use the `terminator` to shut the components down in the reverse order, also sequentially.

Reply:
Implemented the simple approach for now.